contracts/programs/gummyshop/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+        let amount_to_pay_marketplace = (price as f64 * (ctx.accounts.marketplace_props.share as f64/100.0)) as u64;
+        assert_eq!(ctx.accounts.marketplace_receiver.key(), ctx.accounts.marketplace_props.payment_receiver.key());
+        invoke(
+            &system_instruction::transfer(
+                &ctx.accounts.purchaser.key(),
+                &ctx.accounts.marketplace_receiver.key(),
+                amount_to_pay_marketplace
+            ),
+            &[
+                ctx.accounts.purchaser.to_account_info(),
+                ctx.accounts.marketplace_receiver.to_account_info(),
+                ctx.accounts.system_program.to_account_info()
+            ]
+        )?;
+
+        // Second, payout each creator. This should payout 100% of what is left after the marketplace's cut.
+        // Simultaneously, collect creator address, share pairs to prepare to compute creator_hash
+        let total_creator_price_allocation = price - amount_to_pay_marketplace;
+        let mut amount_paid_out_to_creators = 0;
+        let mut creator_data: Vec<Vec<u8>> = Vec::new();
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter();
+        for share in creator_shares.into_iter() {
+            let current_creator_info = next_account_info(remaining_accounts_iter)?;
+            let amount_to_pay_creator = (total_creator_price_allocation as f64 * (share as f64/100.0)) as u64;
+            invoke(
+                &system_instruction::transfer(
+                    &ctx.accounts.purchaser.key(),
+                    &current_creator_info.key(),
+                    amount_to_pay_creator
+                ),
+                &[
+                    ctx.accounts.purchaser.to_account_info(),
+                    current_creator_info.clone(),
+                    ctx.accounts.system_program.to_account_info()
+                ]
+            )?;
+            amount_paid_out_to_creators += amount_to_pay_creator;
+            creator_data.push([current_creator_info.key().as_ref(), &[share]].concat());
+        }
+
+        assert_eq!(amount_paid_out_to_creators, total_creator_price_allocation);
+        
+        // Compute the creator hash using address, share pairs
+        let creator_hash = hashv(
+            creator_data
+                .iter()
+                .map(|c| c.as_slice())
+                .collect::<Vec<&[u8]>>()
+                .as_ref(),
+        );
+
+        // CPI to Bubblegum to transfer the NFT to its new owner
+        let price_seed = price.to_le_bytes();
+        let program_id_seed = id();
+        let seeds = &[price_seed.as_ref(), program_id_seed.as_ref(), &[*ctx.bumps.get("listing_delegate").unwrap()]];
+        let authority_pda_signer = &[&seeds[..]];
+        let cpi_ctx = CpiContext::new_with_signer(
+            ctx.accounts.bubblegum.to_account_info(),
+            bubblegum::cpi::accounts::Transfer {
```

Comment:
My gut tells me that anchor CPI is going to bungle this because there's no ability to specify an optional signer